uwblueprint/website-bp
# Access Token Automatic Refresh

…eperating react from fetch calls


[Handle Automatic Token Refreshing](https://www.notion.so/uwblueprintexecs/58fdb7043f98414aa6f88c7ec827d5ca?v=9df8cc8e794b4d9ea4152c1253b876d6&p=343f22774e044c669148424cac12b228&pm=s)

### Implementation description
- Create BaseAPIClient which should be called by all other api client files to access the handleAuthRefresh. Note that handleAuthRefresh handles automatically refreshing access and refresh tokens which should be stored as cookies under names "accessToken" and "refreshToken"
- Implement AuthAPIClient which handles sending api calls to any auth api endpoints. Only handle isAuthorizedByRole atm
- **NOTE** in the future, all fetch calls should be made via an APIClient class method. In other words, no more raw fetch calls in react components. Instead, we should call some method from an APIClient class.



### Steps to test
Right now, accessTokens default to 60 minutes before expiring. Below, I have a custom video with code edited to refresh the access token regardless of the expiration date. A few notes on what is happening in the video:

1. comment out the code to handle automatic accessToken refresh, forcing the if statement to trigger an accessToken refresh regardless of expiration date
2. go to http://localhost:3000/admin which redirect to http://localhost:3000/admin/login
3. login successfully via google sso which should generate an accessToken in local storage
4. get automatic redirect to http://localhost:3000/admin
5. refresh http://localhost:3000/admin which triggers the react component for ProtectedRoute wrapped around admin page component to re-render. ProtectedRoute retriggers an AuthAPIClient.isAuthorizedByRole call which triggers BaseAPIClient.handleAuthRefresh() that refreshes the accessToken (Normally, BaseAPIClient.handleAuthRefresh() takes 60 minutes before needing to refresh accessToken but our work in 1. allows for automatic accessToken refresh regardless of expiration date.)

Videos below:
https://drive.google.com/file/d/19LIKzLvUFWLNeSiydsy1VMd2_OSnN-nV/view
[Untitled_ Jun 16, 2023 9_18 PM.webm](https://github.com/uwblueprint/website-bp/assets/117593511/d70ea383-5e5b-4687-9567-eaa504561756)



### What should reviewers focus on?
- token access for methods in BaseAPIClient and AuthAPIClient. Since I'm never passing the access or refresh token as an input and the the tokens are instead accessed by querying local storage inside of method bodies, it's important that tokens are being accessed **after they are refreshed**
- Verifying that the accessToken expiration date is properly calculated and the condition for triggering a refresh in handleAuthRefresh is called when accessToken expires. 

### Checklist
